### ISSUE TITLE
Middleware that checks result and rolls back transaction

### ISSUE THREAD

**Issue by mdesousa** [prisma/prisma#15592]
## Problem

Hi, I am looking for a way to implement middleware that can validate the result of an update and potentially rollback the update if validation fails. This would be very useful for validation of business rules when partial updates are done.
An example would be updating the `start_date` column in a table, and checking that `start_date <= `end_date`.

## Suggested solution

I would like to write middleware as shown below. However, it appears that the transaction has completed by the time that the error is thrown. It would be great if the transaction could be rolled back when an error is thrown.

```ts
prisma.$use(async (params, next) => {
  const promise = next(params);
  if (params.action !== 'update') return promise;
  const result = await promise;
  if (result.start_date <= result.end_date) return result;
  throw Error('start_date should happen before end_date');
});
```

## Alternatives

I cannot think of good alternatives to solve this problem... two yucky options:
* query the data before the update executes, and somehow simulate the update to obtain the object that prisma would return... then validate that
* query the data before the update executes. if the validation fails, then update again with the original value.

## Additional context

This would be a tremendously useful feature to ensure that only updates that result in valid data are executed. The start_date <= end_date is a very simple example. The feature would be very useful for many other types of situations:
* for an address, ensure that state is populated only if country = X
* ensuring that a help desk ticket cannot be set to "closed" unless a "solution" field has been populated
* etc...

Thanks!

**Comment by janpio** [comment]
Does this business rule apply to _all_ queries that Prisma Client executes?
Otherwise you might just use an [interactive transaction](https://www.prisma.io/docs/concepts/components/prisma-client/transactions#interactive-transactions-in-preview) instead of a middleware?

**Comment by mdesousa** [comment]
@janpio yes... for all update queries. as background... we use [ajv](https://ajv.js.org/) to validate all our business objects.
so the middleware would take the return value from the update and run the validation (ajv can perform all the validations that i described above).

**Comment by Mark-Panda** [comment]
Is there any progress on this issue? I hope that the instance in the transaction can be audited after the transaction is properly ended, so I hope this function can be realized.

**Comment by Mark-Panda** [comment]
Is there any plan to improve this issue?
